### PR TITLE
Implement 1d convolution via fft

### PR DESCRIPTION
**Pull request by fritzo**
This adds a helper to do long time convolution, as common in signal processing.
The interface matches [np.convolve](https://docs.scipy.org/doc/numpy/reference/generated/numpy.convolve.html) but supports batching and gradients.

## Tested
- added unit tests

### CODE REVIEW

**Comment by fritzo** [comment]
@fehiepsi could you help me out on this PR? I need this helper for a model I'm working on and it's too late at night for me to figure out what's wrong. For reference see [wikipedia](https://en.wikipedia.org/wiki/Convolution_theorem) and the implementation in https://github.com/pytorch/pytorch/issues/21462 (more than I need here).

File: pyro/ops/tensor_utils.py [hunk 1]
```diff
@@ -13,3 +13,30 @@ def block_diag(m):
     B, M, N = m.shape
     eye = torch.eye(B, dtype=m.dtype, device=m.device).reshape(B, 1, B, 1)
     return (m.unsqueeze(-2) * eye).reshape(B * M, B * N)
+
+
+def _complex_mul(a, b):
+    ar, ai = a.unbind(-1)
+    br, bi = b.unbind(-1)
+    # FIXME should b be conjugated?
+    # bi = -bi  # Conjugates b.
```
Comment (fehiepsi) [inline pyro/ops/tensor_utils.py]: I think that no conjugate here.

File: pyro/ops/tensor_utils.py [hunk 2]
```diff
@@ -13,3 +13,30 @@ def block_diag(m):
...
+    arbitrarily broadcastable.
+
+    :param torch.Tensor signal: A signal to convolve.
+    :param torch.Tensor kernel: A convolution kernel.
+    :return: A tensor of shape
+        ``broadcast_shape(convolve, kernel.shape)``
+    :rtype torch.Tensor:
+    """
+    assert signal.size(-1) == kernel.size(-1)
+    f_signal = torch.rfft(torch.cat([signal, torch.zeros_like(signal)], dim=-1), 1)
```
Comment (fehiepsi) [inline pyro/ops/tensor_utils.py]: This is relevant to [autocorrelation](https://github.com/pyro-ppl/pyro/blob/dev/pyro/ops/stats.py#L95). I think [next_fast_len](https://docs.scipy.org/doc/scipy/reference/generated/scipy.fftpack.next_fast_len.html) might be helpful for performance. Let me verify tomorrow if we need `onesided` arg  in `rfft` and `irfft`.

File: pyro/ops/tensor_utils.py [hunk 3]
```diff
@@ -13,3 +13,30 @@ def block_diag(m):
...
+    f_kernel = torch.rfft(torch.cat([kernel, torch.zeros_like(kernel)], dim=-1), 1)
+    f_result = _complex_mul(f_signal, f_kernel)
+    return torch.irfft(f_result, 1)[..., :signal.size(-1)]
```
Comment (fehiepsi) [inline pyro/ops/tensor_utils.py]: I guess you need the output to have the same last dimension as the signal. Probably you will want to take the center of the f_result?

File: tests/ops/test_tensor_utils.py [hunk 4]
```diff
@@ -22,3 +22,18 @@ def test_block_diag(batch_size, block_size):
...
+
+
+@pytest.mark.parametrize('T', [2, 3, 4, 5, 10])
+@pytest.mark.parametrize('batch_shape', [(), (4,), (2, 3)], ids=str)
+def test_conv1d_fft(batch_shape, T):
+    signal = torch.randn(T)  # not batched
+    kernel = torch.randn(*batch_shape, T).exp()  # batched
+    actual = conv1d_fft(signal, kernel)
+
+    # FIXME is this correct?
```
Comment (fehiepsi) [inline tests/ops/test_tensor_utils.py]: I think it is better to compare against https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.fftconvolve.html#scipy.signal.fftconvolve.

File: pyro/ops/tensor_utils.py (hunk 1 above)
Comment (fehiepsi) [reply pyro/ops/tensor_utils.py]: I see, so I think you don't need to take conjugate. :)

File: tests/ops/test_tensor_utils.py (hunk 4 above)
Comment (fritzo) [reply tests/ops/test_tensor_utils.py]: The scipy.signal.convolve should agree with scipy.signal.fftconvolve.

File: pyro/ops/tensor_utils.py [hunk 5]
```diff
@@ -13,3 +15,50 @@ def block_diag(m):
...
+        truncate = m + n - 1
+    elif mode == 'valid':
+        truncate = max(m, n) - min(m, n) + 1
+    elif mode == 'same':
+        truncate = m
+    else:
+        raise ValueError('Unknown mode: {}'.format(mode))
+
+    # Compute convolution using fft.
+    padded_size = m + n
```
Comment (fehiepsi) [inline pyro/ops/tensor_utils.py]: nit: I think `padded_size` would be `m + n - 1` but adding more zeros might have no harm.

File: pyro/ops/tensor_utils.py (hunk 2 above)
Comment (fritzo) [reply pyro/ops/tensor_utils.py]: done, I've updated padding logic.

File: pyro/ops/tensor_utils.py [hunk 6]
```diff
@@ -13,3 +15,50 @@ def block_diag(m):
...
+    :rtype torch.Tensor:
+    """
+    m = signal.size(-1)
+    n = kernel.size(-1)
+    if mode == 'full':
+        truncate = m + n - 1
+    elif mode == 'valid':
+        truncate = max(m, n) - min(m, n) + 1
+    elif mode == 'same':
+        truncate = m
```
[the remark this answers is not in the document]
Comment (fritzo) [reply pyro/ops/tensor_utils.py]: I thing it's best to match scipy.signal.convolve, which defines 'same' as size of the first arg.

File: pyro/ops/tensor_utils.py [hunk 7]
```diff
@@ -13,3 +15,50 @@ def block_diag(m):
...
+    # Compute convolution using fft.
+    padded_size = m + n
+    # Round up to next power of 2 for cheaper fft.
+    padded_size = 2 ** int(math.ceil(math.log2(padded_size)))
+    f_signal = torch.rfft(torch.nn.functional.pad(signal, (0, padded_size - m)), 1)
+    f_kernel = torch.rfft(torch.nn.functional.pad(kernel, (0, padded_size - n)), 1)
+    f_result = _complex_mul(f_signal, f_kernel)
+    result = torch.irfft(f_result, 1)
+
+    return result[..., :truncate]
```
Comment (fehiepsi) [inline pyro/ops/tensor_utils.py]: I think it is better (here I use `padded_size` as m + n - 1, rather than fast_len size) to do
```
result = result[..., :padded_size]
start_idx = (padded_size - truncate) // 2
return result[..., start_idx: start_idx + truncate]

File: pyro/ops/tensor_utils.py (hunk 3 above)
Comment (fritzo) [reply pyro/ops/tensor_utils.py]: The updated truncation logic now matches scipy.signal.convolve in size.

File: tests/ops/test_tensor_utils.py (hunk 4 above)
Comment (fritzo) [reply tests/ops/test_tensor_utils.py]: Actually, np.convolve seems like an even better gold standard.

File: pyro/ops/tensor_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fritzo) [reply pyro/ops/tensor_utils.py]: Oh I see [np.convolve](https://docs.scipy.org/doc/numpy/reference/generated/numpy.convolve.html) follows `truncate = max(m,n)`, so I agree with you. I'll change.

File: pyro/ops/tensor_utils.py [hunk 8]
```diff
@@ -13,3 +15,48 @@ def block_diag(m):
...
+        truncate = max(m, n)
+    else:
+        raise ValueError('Unknown mode: {}'.format(mode))
+
+    # Compute convolution using fft.
+    padded_size = m + n - 1
+    # Round up to next power of 2 for cheaper fft.
+    padded_size = 2 ** int(math.ceil(math.log2(padded_size)))
+    f_signal = torch.rfft(torch.nn.functional.pad(signal, (0, padded_size - m)), 1)
+    f_kernel = torch.rfft(torch.nn.functional.pad(kernel, (0, padded_size - n)), 1)
```
Comment (fehiepsi) [inline pyro/ops/tensor_utils.py]: ditto: `onesided=False`

File: pyro/ops/tensor_utils.py [hunk 9]
```diff
@@ -13,3 +15,48 @@ def block_diag(m):
...
+    f_result = _complex_mul(f_signal, f_kernel)
+    result = torch.irfft(f_result, 1)
```
Comment (fehiepsi) [inline pyro/ops/tensor_utils.py]: ditto: `onesided=False`

File: pyro/ops/tensor_utils.py [hunk 10]
```diff
@@ -13,3 +15,48 @@ def block_diag(m):
...
+        truncate = max(m, n) - min(m, n) + 1
+    elif mode == 'same':
+        truncate = max(m, n)
+    else:
+        raise ValueError('Unknown mode: {}'.format(mode))
+
+    # Compute convolution using fft.
+    padded_size = m + n - 1
+    # Round up to next power of 2 for cheaper fft.
+    padded_size = 2 ** int(math.ceil(math.log2(padded_size)))
```
Comment (fehiepsi) [inline pyro/ops/tensor_utils.py]: nit: no need for `int`.

**Comment by fritzo** [comment]
Thanks for all the help @fehiepsi !

**Comment by ctr26** [comment]
Any thoughts on this being implemented directly into pyTorch?

**Comment by fritzo** [comment]
@ctr26 you're welcome to move this upstream to PyTorch. I believe they would welcome any contribution that closely matches the NumPy interface. When we're busy we often simply implement things in Pyro and move on.
